atecc608a_se.c

```
@@ -0,0 +1,212 @@
+    case ATCA_PARSE_ERROR:
+    case ATCA_FUNC_FAIL:
+    case ATCA_GEN_FAIL:
+    case ATCA_EXECUTION_ERROR:
+    case ATCA_HEALTH_TEST_ERROR:
+    default:
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+}
+
+psa_status_t atecc608a_get_serial_number(uint8_t* buffer, size_t buffer_size)
+{
+
+    if (buffer_size < ATCA_SERIAL_NUM_SIZE)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_read_serial_number(buffer) != ATCA_SUCCESS)
+    {
+        atcab_release();
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_release() != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atecc608a_check_config_locked()
+{
+    bool config_locked;
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_is_locked(LOCK_ZONE_CONFIG, &config_locked) != ATCA_SUCCESS)
+    {
+        atcab_release();
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if (atcab_release() != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    return config_locked? PSA_SUCCESS : PSA_ERROR_HARDWARE_FAILURE;
+}
+
+psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data, size_t data_size,
+                                         size_t *p_data_length)
+{
+    const size_t key_data_len = 65;
+    const uint16_t slot = key;
+    ATCA_STATUS ret = ATCA_SUCCESS;
+
+    if (data_size < key_data_len)
+    {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (atcab_init(&atca_iface_config) != ATCA_SUCCESS)
+    {
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+
+    if ((ret = atcab_get_pubkey(slot, &p_data[1])) != ATCA_SUCCESS)
+    {
+        atcab_release();
```

If `actab_release` fails, that means there's something wrong with the hardware or there's a driver bug. Should we return `HARDWARE_FAILURE` in that case? Mbed Crypto currently tends to ignore cleanup errors when an error has already happened, but I wonder if that's the right decision. If something is wrong during cleanup, like failing hardware or memory corruption, shouldn't we let the caller know?